pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,318 @@
+	deprovisioningevents "github.com/aws/karpenter-core/pkg/controllers/deprovisioning/events"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/utils/nodeclaim"
+)
+
+const (
+	queueBaseDelay   = 1 * time.Second
+	queueMaxDelay    = 10 * time.Second
+	maxRetryDuration = 10 * time.Minute
+)
+
+type Command struct {
+	ReplacementKeys []*NodeClaimKey
+	Candidates      []*state.StateNode
+	Reason          string    // reason is used for metrics
+	TimeAdded       time.Time // timeAdded is used to track timeouts
+	LastError       error
+}
+
+// NodeClaimKey wraps a nodeclaim.Key with an initialized field to save on readiness checks and identify
+// when a nodeclaim is first initialized for metrics and events.
+type NodeClaimKey struct {
+	nodeclaim.Key
+	// Use a bool track if a node has already been initialized. This intentionally does not capture nodes that go
+	// initialized then go NotReady after as other pods can schedule to this node as well.
+	Initialized bool
+}
+
+type Queue struct {
+	workqueue.RateLimitingInterface
+	// providerID -> command, maps a candidate to its command. Each command has a list of candidates that can be used
+	// to map to itself.
+	candidateProviderIDToCommand map[string]*Command
+
+	kubeClient client.Client
+	recorder   events.Recorder
+	cluster    *state.Cluster
+	clock      clock.Clock
+}
+
+// NewQueue creates a queue that will asynchronously orchestrate deprovisioning commands
+func NewQueue(ctx context.Context, kubeClient client.Client, recorder events.Recorder, cluster *state.Cluster, clock clock.Clock, testingMode bool) *Queue {
+	queue := &Queue{
+		RateLimitingInterface:        workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
+		candidateProviderIDToCommand: map[string]*Command{},
+		kubeClient:                   kubeClient,
+		recorder:                     recorder,
+		cluster:                      cluster,
+		clock:                        clock,
+	}
+	// If testing, we don't actually need to start the queue. We can call each function individually instead of relying
+	// on the rate limiting interface itself.
+	if !testingMode {
+		go queue.Start(logging.WithLogger(ctx, logging.FromContext(ctx).Named("deprovisioning")))
+	}
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, reason string, timeAdded time.Time) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates: candidates,
+		Reason:     reason,
+		TimeAdded:  timeAdded,
+	}
+}
+
+// Start will kick off the queue. This runs asynchronously until Karpenter shuts down.
+func (q *Queue) Start(ctx context.Context) {
```

Comment:
Consider moving this to the constructor. Is this public method being used elsewhere?